File: script/audio.js
define('audio', ['buzz'], function(buzz) {
    'use strict';

    var loops = {};
    var muted = false;
    var playingLoop = null;

    function loadLoop(name, url) {
        if (name in loops) return;
        loops[name] = new buzz.sound(
            url,
            {
                formats: ['mp3'],
                preload: true,
                autoload: true,
                loop: true
            }
        );
    }

    loadLoop('title', 'audio/title');
    loadLoop('hero', 'audio/hero');

    window.addEventListener('beforeunload', function() {
        if (muted) {
            window.localStorage.muted = '1';
        } else {
            delete window.localStorage.muted;
        }
    });

    function toggleMute() {
        muted = !muted;
        Object.keys(loops).forEach(function(loop) {
            loops[loop].toggleMute();
        });
    }

    if ('muted' in window.localStorage) {
        toggleMute();
    }

    return {
        loadLoop: loadLoop,
        playLoop: function(name) {
            if (playingLoop == name)
                return;

            if (!playingLoop) {
                loops[name].play().setVolume(0).fadeTo(10, 1000);
                playingLoop = name;
                return;
            }

            loops[playingLoop].fadeOut(2000, function() {
                playingLoop = name;
                // FIXME: Bad things might happen if playLoop is called again
                // within four seconds of it being called once.
                loops[name].play().setVolume(0).fadeTo(20, 2000);
            });
        },
        isMuted: function() {
            return muted;
        },
        toggleMute: toggleMute,
        stop: function() {
            if (!playingLoop) return;
            loops[playingLoop].stop();
            playingLoop = null;
        }
    };
});

File: script/audio.dart
import 'dart:js';


Map<String, JsObject> loops = {};
String playingLoop = null;

void _loadLoop(String name, String uri) {
    if (loops.containsKey(name)) {
        return;
    }

    var loop = new JsObject(
        context['buzz']['sound'],
        [
            uri,
            new JsObject.jsify({'formats': ['mp3'], 'preload': true, 'autoload': true, 'loop': true}),
        ]
    );
    loops[name] = loop;
}

void init() {
    _loadLoop('title', 'audio/title');
    _loadLoop('hero', 'audio/hero');
}

void playLoop(String name) {
    if (playingLoop == name) {
        return;
    }

    if (playingLoop == null) {
        loops[name]..callMethod('play')
                   ..callMethod('setVolume', [0])
                   ..callMethod('fadeTo', [10, 1000]);
        playingLoop = name;
        return;
    }

    loops[playingLoop].callMethod('fadeOut', [2000, () {
        playingLoop = name;
        loops[name]..callMethod('play')
                   ..callMethod('setVolume', [0])
                   ..callMethod('fadeTo', [20, 2000]);
    }]);
}

void stop() {
    if (playingLoop == null) {
        return;
    }
    loops[playingLoop].callMethod('stop');
    playingLoop = null;
}
